Makefile

```diff
@@ -112,59 +115,94 @@ demo: ## Execute end-to-end demo
 
 # spin-up local environment
 .PHONY: deploy-full-local-setup
-deploy-full-local-setup: ## Deploy full local multicluster setup (k3d >= 4.2.0)
-	$(call create-local-cluster,$(CLUSTER_GSLB1))
-	$(call create-local-cluster,$(CLUSTER_GSLB2))
+deploy-full-local-setup: ensure-cluster-size ## Deploy full local multicluster setup (k3d >= 4.2.0)
+	@echo "\n$(YELLOW)Creating $(CLUSTERS_NUMBER) k8s clusters$(NC)"
+	@for c in $(CLUSTER_IDS); do \
+		$(MAKE) create-local-cluster CLUSTER_NAME=$(CLUSTER_NAME)$$c ;\
+	done
 
-	$(call deploy-local-cluster,$(CLUSTER_GSLB1),$(CLUSTER_GSLB2),$(VERSION),,'k8gb/k8gb')
-	$(call deploy-local-cluster,$(CLUSTER_GSLB2),$(CLUSTER_GSLB1),$(VERSION),$(CLUSTER_GSLB2_HELM_ARGS),'k8gb/k8gb')
-
-	@echo "\n$(YELLOW)Deploy test apps $(NC)"
-	$(call deploy-test-apps,$(CLUSTER_GSLB1))
-	$(call deploy-test-apps,$(CLUSTER_GSLB2))
+	$(MAKE) deploy-stable-version
 
 .PHONY: deploy-stable-version
 deploy-stable-version:
-	@echo "\n$(YELLOW) import $(CYAN)k8gb:$(STABLE_VERSION) $(YELLOW)to $(CYAN)$(CLUSTER_GSLB1), $(CLUSTER_GSLB2) $(NC)"
-	$(call deploy-local-cluster,$(CLUSTER_GSLB1),$(CLUSTER_GSLB2),$(STABLE_VERSION),,'k8gb/k8gb')
-	$(call deploy-local-cluster,$(CLUSTER_GSLB2),$(CLUSTER_GSLB1),$(STABLE_VERSION),$(CLUSTER_GSLB2_HELM_ARGS),'k8gb/k8gb')
-
-	$(call list-running-pods,$(CLUSTER_GSLB1))
-	$(call list-running-pods,$(CLUSTER_GSLB2))
+	@for c in $(CLUSTER_IDS); do \
+		$(MAKE) deploy-local-cluster CLUSTER_ID=$$c ;\
+	done
 
 .PHONY: deploy-test-version
 deploy-test-version: ## Upgrade k8gb to the test version on existing clusters
-	@echo "\n$(YELLOW)import k8gb docker image to $(CYAN)$(CLUSTER_GSLB1), $(CLUSTER_GSLB2) $(NC)"
+	@echo "\n$(YELLOW)import k8gb docker image to all $(CLUSTERS_NUMBER) clusters$(NC)"
+
+	@for c in $(CLUSTER_IDS); do \
+		echo "\n$(CYAN)$(CLUSTER_NAME)$$c:$(NC)" ;\
+		k3d image import $(REPO):$(SEMVER)-amd64 -c $(CLUSTER_NAME)$$c ;\
+	done
+
+	@for c in $(CLUSTER_IDS); do \
+		$(MAKE) deploy-local-cluster CLUSTER_ID=$$c VERSION=$(SEMVER)-amd64 CHART='./chart/k8gb' ;\
```

Comment:
IIRC, we've tried to get rid of that practice of calling external `make` as it's hard to debug and decided to use custom functions as means of code reuse. Is there something preventing us from following this approach?

Reply:
yes, good point, I didn't know that you were actually doing it before, couple of reasons:

1) I think it's actually easier to debug, because one can call `make foobar` directly from terminal with arbitrary parameters w/o having to introduce special makefile target just to verify if the function/macro works

2) named arguments vs positional arguments, good example is this:
  ```bash
  $(call deploy-k8gb-with-helm,$(CLUSTER_GSLB1),$(CLUSTER_GSLB2),$(SEMVER)-amd64,,'./chart/k8gb')
  ```
  
  vs just (as it's called in the PR)
  ```bash
  $(MAKE) deploy-k8gb-with-helm
  ```
  ok little bit of cheating, because stuff got passed implicitly as vars, but thats good actually.. no need to pass everything over and over again, only those things that are different than defaults. Check like 129 for instance, here the `CHART` is not "overiden" and the default (`k8gb/k8gb`) is taken.

3) most importantly the `$(call foobar,arg1,arg2)` do not work reliably in for loops. If makefile contains for loop or a conditional, the commands in the body need to be separated by some non-whitespace separator (you know, `;`, `&&`,`||`, etc.) but on the other hand when commands are listed one by one under makefile target, each command is executed in its own subshell and those commands are not separated.
With these two different context we have no reliable way to call the macro from both, the loop context (requires separator) and from the regular target line (no separator). There are [workarounds](https://stackoverflow.com/a/12605404) and I went down that route, but it would complicate stuff even more.

4) crossplane/upbound also do that and I found their Makefile machinery pretty good and easy to use, I think we should actually split our huge Makefile into smaller ones (one with functions only, one for CI perhaps, one for local development, observability targets can live on their own etc.) but that's different story. 
My first interaction with k8gb as an external guy was, after cloning the repo, going to the Makefile and trying to figure out what that `make deploy-full-local-setup` or `make test-round-robin` do and sometimes it would be actually use the kubectl command directly instead of hiding it behind Makefile target (introducing unnecessary level of indirection).
I can imagine how it was done, that you wanted to have the Makefile DRY so that it was encapsulated as a funciton and then referrenced from the local development tutorial, but I find much easier to follow what this does
  ```bash
  kubectl scale deployment frontend-podinfo -n test-gslb --replicas=0
  ```
  ..rather than inspecting Makefile and learning that `make start-test-app` -> `$(call testapp-set-replicas,0)` -> `kubectl scale deploy ..`. 
  btw. I didn't touch this part at all

Reply:
> 1. I think it's actually easier to debug, because one can call make foobar directly from terminal with arbitrary parameters w/o having to introduce special makefile target just to verify if the function/macro works
You can usually run `make -d` or `make --debug=FLAGS` to generate advanced debug tracing during make target run. When you reinvoke `make` from make targets, you are losing this setting (and tracing as result) unless you explicitly pass that option again.

Reply:
> most importantly the $(call foobar,arg1,arg2) do not work reliably in for loops. If makefile contains for loop or a conditional, the commands in the body need to be separated by some non-whitespace separator (you know, ;, &&,||, etc.) but on the other hand when commands are listed one by one under makefile target, each command is executed in its own subshell and those commands are not separated.

just a thought: can make's `foreach` function help here? https://www.gnu.org/software/make/manual/html_node/Foreach-Function.html#Foreach-Function

Reply:
> ok little bit of cheating, because stuff got passed implicitly as vars, but thats good actually.. no need to pass everything over and over again, only those things that are different than defaults. 

IMHO explicit is better than implicit magic in this case, because env vars are globals and it's hard to trace sometimes what implicit value they have at the moment of call.

Reply:
> You can usually run make -d or make --debug=FLAGS...

These actually propagate correctly to recursive calls, check:
Makefile:
```makefile
foo:
	echo "hello from foo"
	echo $(MAKEFLAGS)
	$(MAKE) bar

bar:
	echo "hello from bar"
	echo $(MAKEFLAGS)
```

`make -d foo`
both targets will be run w/ that `-d` option and both will echo `-d`
Some options are not [propagated](https://www.gnu.org/software/make/manual/html_node/Options_002fRecursion.html#Options_002fRecursion):
> "The options ‘-C’, ‘-f’, ‘-o’, and ‘-W’ are not put into MAKEFLAGS; these options are not passed down."

but this is pretty edge case to me and it can be controlled by altering that `MAKEFLAGS` var

Reply:
>  can make's foreach function help here?

it can, I've actually had a link to the [workaround](https://stackoverflow.com/a/12605404) in the original comment and I was doing it at the beginning, but it was too complicated wrapping it in this one-liner. Also only 1 type of command separator can be used here